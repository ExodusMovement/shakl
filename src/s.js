import React from 'react';

import flatten from './flatten';

const s = (Comp, { name, multi, ...opts } = {}) => style => {
  const S = React.forwardRef(({ children, ...props }, ref) => {
    const { attrs = {}, comp, child } = opts;

    const styles = multi ? style : { style };

    return React.createElement(
      comp || Comp,
      {
        ref,
        ...attrs,
        ...props,
        ...Object.keys(styles).reduce(
          (obj, prop) => ({
            ...obj,
            [prop]: flatten([
              attrs[prop],
              typeof styles[prop] === 'function'
                ? styles[prop](props)
                : styles[prop],
              props[prop]
            ])
          }),
          {}
        )
      },
      child ? React.createElement(child, {}, children) : children
    );
  });

  S.displayName = name || `styled(${Comp.displayName || Comp.name})`;

  S.extend = more => s(S, { name })(more);
  S.withProps = attrs => s(S, { name, attrs })();
  S.withComponent = comp => s(S, { name, comp })(style);
  S.withChild = child => s(S, { name, child })();

  return S;
};

export default s;
