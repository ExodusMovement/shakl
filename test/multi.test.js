import React from 'react';

import { create as r } from 'react-test-renderer';

import { FlatList } from 'react-native';

import s from '../src';

it('creates a styled component with static multi style props', () => {
  const Foo = s(FlatList, { multi: true })({
    style: { flex: 1 },
    contentContainerStyle: { flex: 2 },
    anotherStyleProp: { flex: 3 }
  });

  const foo = r(<Foo />).toJSON();

  expect(foo.props.style).toEqual({ flex: 1 });
  expect(foo.props.contentContainerStyle).toEqual({ flex: 2 });
  expect(foo.props.anotherStyleProp).toEqual({ flex: 3 });

  expect(foo).toMatchSnapshot();
});

it('creates a styled component with dynamic multi style props', () => {
  const Foo = s(FlatList, { multi: true })({
    style: ({ padded }) => ({ padding: padded ? 10 : 0 }),
    contentContainerStyle: ({ padded }) => ({ padding: padded ? 20 : 0 }),
    anotherStyleProp: ({ padded }) => ({ padding: padded ? 30 : 0 })
  });

  const foo = r(<Foo />).toJSON();
  const fooPadded = r(<Foo padded />).toJSON();

  expect(foo.props.style).toEqual({ padding: 0 });
  expect(foo.props.contentContainerStyle).toEqual({ padding: 0 });
  expect(foo.props.anotherStyleProp).toEqual({ padding: 0 });
  expect(fooPadded.props.style).toEqual({ padding: 10 });
  expect(fooPadded.props.contentContainerStyle).toEqual({ padding: 20 });
  expect(fooPadded.props.anotherStyleProp).toEqual({ padding: 30 });

  expect(foo).toMatchSnapshot();
  expect(fooPadded).toMatchSnapshot();
});
