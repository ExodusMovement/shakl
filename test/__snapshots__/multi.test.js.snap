// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a styled component with dynamic multi style props 1`] = `
<RCTScrollView
  anotherStyleProp={
    {
      "padding": 0,
    }
  }
  contentContainerStyle={
    {
      "padding": 0,
    }
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  style={
    {
      "padding": 0,
    }
  }
  viewabilityConfigCallbackPairs={[]}
>
  <View />
</RCTScrollView>
`;

exports[`creates a styled component with dynamic multi style props 2`] = `
<RCTScrollView
  anotherStyleProp={
    {
      "padding": 30,
    }
  }
  contentContainerStyle={
    {
      "padding": 20,
    }
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  padded={true}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  style={
    {
      "padding": 10,
    }
  }
  viewabilityConfigCallbackPairs={[]}
>
  <View />
</RCTScrollView>
`;

exports[`creates a styled component with static multi style props 1`] = `
<RCTScrollView
  anotherStyleProp={
    {
      "flex": 3,
    }
  }
  contentContainerStyle={
    {
      "flex": 2,
    }
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  style={
    {
      "flex": 1,
    }
  }
  viewabilityConfigCallbackPairs={[]}
>
  <View />
</RCTScrollView>
`;
